modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/upgrade/v4_3_1/test/DDMFormInstanceUpgradeProcessTest.java

```diff
@@ -0,0 +1,121 @@
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.dynamic.data.mapping.upgrade.v4_3_1.test;
+
+import com.liferay.arquillian.extension.junit.bridge.junit.Arquillian;
+import com.liferay.change.tracking.test.util.BaseCTUpgradeProcessTestCase;
+import com.liferay.dynamic.data.mapping.model.DDMForm;
+import com.liferay.dynamic.data.mapping.model.DDMFormInstance;
+import com.liferay.dynamic.data.mapping.service.DDMFormInstanceLocalService;
+import com.liferay.dynamic.data.mapping.storage.DDMFormValues;
+import com.liferay.dynamic.data.mapping.test.util.DDMFormInstanceTestUtil;
+import com.liferay.dynamic.data.mapping.test.util.DDMFormTestUtil;
+import com.liferay.dynamic.data.mapping.test.util.DDMFormValuesTestUtil;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.model.Group;
+import com.liferay.portal.kernel.model.change.tracking.CTModel;
+import com.liferay.portal.kernel.service.change.tracking.CTService;
+import com.liferay.portal.kernel.test.rule.AggregateTestRule;
+import com.liferay.portal.kernel.test.rule.DeleteAfterTestRun;
+import com.liferay.portal.kernel.test.util.GroupTestUtil;
+import com.liferay.portal.kernel.test.util.RandomTestUtil;
+import com.liferay.portal.kernel.test.util.TestPropsValues;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+import com.liferay.portal.test.rule.PermissionCheckerMethodTestRule;
+import com.liferay.portal.upgrade.registry.UpgradeStepRegistrator;
+import com.liferay.portal.upgrade.test.util.UpgradeTestUtil;
+
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.runner.RunWith;
+
+/**
+ * @author Paulo Albuquerque
+ */
+@RunWith(Arquillian.class)
+public class DDMFormInstanceUpgradeProcessTest
+	extends BaseCTUpgradeProcessTestCase {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new AggregateTestRule(
+			new LiferayIntegrationTestRule(),
+			PermissionCheckerMethodTestRule.INSTANCE);
+
+	@Before
+	public void setUp() throws Exception {
+		_group = GroupTestUtil.addGroup();
+	}
+
+	@Override
+	protected CTModel<?> addCTModel() throws Exception {
+		DDMForm ddmForm = DDMFormTestUtil.createDDMForm(
+			RandomTestUtil.randomString());
+
+		DDMFormValues ddmFormValues = new DDMFormValues(ddmForm);
+
+		ddmFormValues.setAvailableLocales(ddmForm.getAvailableLocales());
+		ddmFormValues.setDefaultLocale(ddmForm.getDefaultLocale());
+
+		ddmFormValues.addDDMFormFieldValue(
+			DDMFormValuesTestUtil.createUnlocalizedDDMFormFieldValue(
+				"workflowDefinition",
+				StringBundler.concat(
+					"[\"", RandomTestUtil.randomString(), "@",
+					RandomTestUtil.randomInt(), "\"]")));
+
+		_ddmFormInstance = DDMFormInstanceTestUtil.addDDMFormInstance(
+			ddmForm, _group, ddmFormValues, TestPropsValues.getUserId());
+
```

Comment:
@p-albuquerque I noticed that you are defining the `ddmFormValues` for the `ddmFormInstance` **settings**, so it makes more sense to do something like:

```
DDMForm ddmForm = DDMFormFactory.create(DDMFormInstanceSettings.class);

DDMFormValues ddmFormValues = DDMFormValuesTestUtil.createDDMFormValues(ddmForm);

ddmFormValues.addDDMFormFieldValue(
	DDMFormValuesTestUtil.createUnlocalizedDDMFormFieldValue(
		"workflowDefinition", ....);

_ddmFormInstance = DDMFormInstanceTestUtil.addDDMFormInstance(
	DDMFormTestUtil.createDDMForm(RandomTestUtil.randomString()), 
	_group, ddmFormValues, TestPropsValues.getUserId());
```

Reply:
@carolmariaabb Do you mean:

```java
DDMFormValues ddmFormValues = DDMFormValuesTestUtil.createDDMFormValues(
	DDMFormFactory.create(DDMFormInstanceSettings.class));

ddmFormValues.addDDMFormFieldValue(
	DDMFormValuesTestUtil.createUnlocalizedDDMFormFieldValue(
		"workflowDefinition",
		...));

_ddmFormInstance = DDMFormInstanceTestUtil.addDDMFormInstance(
	DDMFormTestUtil.createDDMForm(RandomTestUtil.randomString()),
	_group, ddmFormValues, TestPropsValues.getUserId());
```

Reply:
@p-albuquerque yes